tests/Unit/ActionTesting.hpp

```diff
@@ -926,4 +946,62 @@ class MockRuntimeSystem {
   Inboxes inboxes_;
   TupleOfMockDistributedObjects local_algorithms_;
 };
+
+/// Returns a vector of all the indices of the Components
+/// in the ComponentList that have queued actions.
+template <typename ComponentList, typename MockRuntimeSystem>
+std::vector<size_t> indices_of_components_with_queued_actions(
+    MockRuntimeSystem& runner) noexcept {
+  std::vector<size_t> result{};
+  size_t i = 0;
+  tmpl::for_each<ComponentList>([&](auto tag) noexcept {
+    using Tag = typename decltype(tag)::type;
+    if (not runner.template is_simple_action_queue_empty<Tag>(0)) {
+      result.push_back(i);
+    }
+    ++i;
+  });
+  return result;
+}
+
+/// \cond
+namespace detail {
+// Helper struct that invokes the next queued action on
+// the ith Component.
+template <typename ComponentList,
```

Comment:
You should be able to do this without recursion by taking a `std::index_sequence<Is...>` and then using `EXPAND_PACK_LEFT_TO_RIGHT(helper_lambda(std::integral_constant<size_t, Is>{})...);`
```cpp
template <class ComponentList, size_t... Is>
void invoke_queued_action(const gsl::not_null<MockRuntimeSystem*> runner,
      const size_t component_to_invoke, std::index_sequence<Is...> /*meta*/) noexcept {
  const auto helper_lambda = [component_to_invoke, &runner](auto I) noexcept {
    if (I.value == component_to_invoke) {
      // I'm a bit concerned about passed `0` since that assumes the ID type is implicitly
      // constructible from a size_t. We probably want to invoke an all members of the
      // component anyway.
      runner->template invoke_queued_simple_action<tmpl::at_c<ComponentList, I::value>>();
    }
  };
  EXPAND_PACK_LEFT_TO_RIGHT(helper_lambda(std::integral_constant<size_t, Is>{})...);
}
```